File: src/main/java/com/codingforcookies/betterrecords/block/tile/TileSpeaker.java
package com.codingforcookies.betterrecords.block.tile;

import com.codingforcookies.betterrecords.api.connection.RecordConnection;
import com.codingforcookies.betterrecords.api.wire.IRecordWire;
import com.codingforcookies.betterrecords.common.core.helper.ConnectionHelper;
import net.minecraft.nbt.NBTTagCompound;

import java.util.ArrayList;

public class TileSpeaker extends BetterTile implements IRecordWire {
    public ArrayList<RecordConnection> connections = null;
    public ArrayList<RecordConnection> getConnections() { return connections; }

    private static String[] prefix = new String[]{ "Small", "Medium", "Large" };
    private static float[] radius = new float[]{ 15F, 35F, 70F };

    public String getName() { return type == -1 ? "Invalid Speaker" : prefix[type] + " Speaker"; }
    public float getSongRadiusIncrease() { return type == -1 ? 0 : radius[type]; }

    public int type = -1;
    public float rotation = 0F;

    public TileSpeaker() {
        connections = new ArrayList<RecordConnection>();
    }

    public boolean canUpdate() {
        return false;
    }

    @Override
    public void readFromNBT(NBTTagCompound compound) {
        super.readFromNBT(compound);

        if(compound.hasKey("type"))
            type = compound.getInteger("type");
        if(compound.hasKey("rotation"))
            rotation = compound.getInteger("rotation");
        if(compound.hasKey("connections"))
            connections = ConnectionHelper.unserializeConnections(compound.getString("connections"));
    }

    @Override
    public NBTTagCompound writeToNBT(NBTTagCompound compound) {
        compound = super.writeToNBT(compound);

        compound.setInteger("type", type);
        compound.setFloat("rotation", rotation);
        compound.setString("connections", ConnectionHelper.serializeConnections(connections));

        return compound;
    }
}

File: src/main/kotlin/com/codingforcookies/betterrecords/block/tile/TileSpeaker.kt
package com.codingforcookies.betterrecords.block.tile

import com.codingforcookies.betterrecords.api.connection.RecordConnection
import com.codingforcookies.betterrecords.api.wire.IRecordWire
import com.codingforcookies.betterrecords.block.BlockSpeaker
import com.codingforcookies.betterrecords.helper.ConnectionHelper
import com.codingforcookies.betterrecords.extensions.set
import net.minecraft.nbt.NBTTagCompound

class TileSpeaker : ModTile(), IRecordWire {

    var rotation = 0f

    var size = BlockSpeaker.SpeakerSize.SMALL
        get() = world?.getBlockState(pos)?.getValue(BlockSpeaker.PROPERTYSIZE) ?: BlockSpeaker.SpeakerSize.SMALL


    override var connections = mutableListOf<RecordConnection>()

    override fun getName() = when (size) {
        BlockSpeaker.SpeakerSize.SMALL -> "Small"
        BlockSpeaker.SpeakerSize.MEDIUM -> "Medium"
        BlockSpeaker.SpeakerSize.LARGE -> "Large"
    } + " Speaker"

    override val songRadiusIncrease
        get() = when (size) {
            BlockSpeaker.SpeakerSize.SMALL -> 15F
            BlockSpeaker.SpeakerSize.MEDIUM -> 35F
            BlockSpeaker.SpeakerSize.LARGE -> 70F
        }

    override fun readFromNBT(compound: NBTTagCompound) = compound.run {
        super.readFromNBT(compound)

        rotation = getFloat("rotation")
        connections = ConnectionHelper.unserializeConnections(getString("connections")).toMutableList()
    }

    override fun writeToNBT(compound: NBTTagCompound) = compound.apply {
        super.writeToNBT(compound)

        set("rotation", rotation)
        set("connections", ConnectionHelper.serializeConnections(connections))
    }

}
